porch/pkg/engine/engine.go

```diff
@@ -846,86 +846,103 @@ func (cad *cadEngine) DeletePackage(ctx context.Context, repositoryObj *configap
+		return nil, nil, fmt.Errorf("cannot update a package revision with lifecycle value %q; package must be Draft", lifecycle)
 	}
 
 	repo, err := cad.cache.OpenRepository(ctx, repositoryObj)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-
 	draft, err := repo.UpdatePackageRevision(ctx, oldPackage.repoPackageRevision)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	mutations := []mutation{
 		&mutationReplaceResources{
 			newResources: new,
 			oldResources: old,
 		},
-		&renderPackageMutation{
-			renderer: cad.renderer,
-			runtime:  cad.runtime,
-		},
 	}
-
-	apiResources, err := oldPackage.repoPackageRevision.GetResources(ctx)
+	prevResources, err := oldPackage.repoPackageRevision.GetResources(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get package resources: %w", err)
+		return nil, nil, fmt.Errorf("cannot get package resources: %w", err)
 	}
 	resources := repository.PackageResources{
-		Contents: apiResources.Spec.Resources,
+		Contents: prevResources.Spec.Resources,
+	}
+	appliedResources, _, err := applyResourceMutations(ctx, draft, resources, mutations)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	if err := applyResourceMutations(ctx, draft, resources, mutations); err != nil {
-		return nil, err
+	// render the package
+	_, renderStatus, err := applyResourceMutations(ctx,
+		draft,
+		appliedResources,
+		[]mutation{&renderPackageMutation{
+			renderer: cad.renderer,
+			runtime:  cad.runtime,
+		}})
+	if err != nil {
+		// Render failure will not result in the overall API operation.
+		// The render error (and results) is also captured as part of renderStatus above
+		// that is saved in packageresourceresources API's status field. We continue with
+		// saving the non-rendered resources to avoid losing user's changes.
+		// and supress this err
+		err = nil
 	}
 
 	// No lifecycle change when updating package resources; updates are done.
 	repoPkgRev, err := draft.Close(ctx)
 	if err != nil {
-		return nil, err
+		return nil, renderStatus, err
 	}
 	return &PackageRevision{
 		repoPackageRevision: repoPkgRev,
-	}, nil
+	}, renderStatus, nil
 }
 
-func applyResourceMutations(ctx context.Context, draft repository.PackageDraft, baseResources repository.PackageResources, mutations []mutation) error {
+// applyResourceMutations mutates the resources and returns the most recent renderResult.
+func applyResourceMutations(ctx context.Context, draft repository.PackageDraft, baseResources repository.PackageResources, mutations []mutation) (applied repository.PackageResources, renderStatus *api.RenderStatus, err error) {
 	for _, m := range mutations {
-		applied, task, err := m.Apply(ctx, baseResources)
+		updatedResources, taskResult, task, err := m.Apply(ctx, baseResources)
+		if taskResult != nil && taskResult.Type == api.TaskTypeEval {
+			renderStatus = taskResult.RenderStatus
```

Comment:
Does this mean we only capture the last `eval` (i.e. the last function?)

Reply:
Yes.

Reply:
But is the goal that it should contain the output/errors from every function in the pipeline? Or do we abort the run when a function return an error?